src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -5244,12 +5251,22 @@ void StorageReplicatedMergeTree::truncate(
 
     Strings partitions = zookeeper->getChildren(fs::path(zookeeper_path) / "block_numbers");
 
+    std::vector<std::unique_ptr<LogEntry>> entries_to_wait;
+    entries_to_wait.reserve(partitions.size());
     for (String & partition_id : partitions)
     {
-        LogEntry entry;
+        auto entry = std::make_unique<LogEntry>();
+        if (dropAllPartsInPartition(*zookeeper, partition_id, *entry, query_context, false))
+            entries_to_wait.push_back(std::move(entry));
+    }
 
-        if (dropAllPartsInPartition(*zookeeper, partition_id, entry, query_context, false))
-            waitForAllReplicasToProcessLogEntry(entry);
+    Int64 wait_for_inactive_timeout = query_context->getSettingsRef().replication_wait_for_inactive_replica_timeout;
+    for (const auto & entry : entries_to_wait)
+    {
+        if (query_context->getSettingsRef().replication_alter_partitions_sync == 1)
+            waitForReplicaToProcessLogEntry(replica_name, *entry, wait_for_inactive_timeout);
```

Comment:
As far as I can see this won't throw if it's waiting on itself and it times out (not a shutdown). Should it? Maybe it should use some `tryWaitForReplicaToProcessLogEntry` instead?

Reply:
I think it should always throw if entry was not processed on specified replica[s] for any reason.